packages/framework/ai-collab/src/explicit-strategy/agentEditReducer.ts

```diff
@@ -0,0 +1,499 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { assert } from "@fluidframework/core-utils/internal";
+import { isFluidHandle } from "@fluidframework/runtime-utils";
+import { UsageError } from "@fluidframework/telemetry-utils/internal";
+import {
+	Tree,
+	NodeKind,
+	type ImplicitAllowedTypes,
+	type TreeArrayNode,
+	type TreeNode,
+	type TreeNodeSchema,
+	type SimpleNodeSchema,
+	FieldKind,
+	FieldSchema,
+	normalizeAllowedTypes,
+	type ImplicitFieldSchema,
+	booleanSchema,
+	handleSchema,
+	nullSchema,
+	numberSchema,
+	stringSchema,
+} from "@fluidframework/tree/internal";
+
+import {
+	type TreeEdit,
+	type ObjectTarget,
+	type Selection,
+	type Range,
+	type ObjectPlace,
+	type ArrayPlace,
+	type TreeEditObject,
+	type TreeEditValue,
+	typeField,
+} from "./agentEditTypes.js";
+import type { IdGenerator } from "./idGenerator.js";
+import type { JsonValue } from "./jsonTypes.js";
+import { toDecoratedJson } from "./promptGeneration.js";
+import { fail } from "./utils.js";
+
+function populateDefaults(
+	json: JsonValue,
+	definitionMap: ReadonlyMap<string, SimpleNodeSchema>,
+): void {
+	if (typeof json === "object") {
+		if (json === null) {
+			return;
+		}
+		if (Array.isArray(json)) {
+			for (const element of json) {
+				populateDefaults(element, definitionMap);
+			}
+		} else {
+			assert(
+				typeof json[typeField] === "string",
+				"The typeField must be present in new JSON content",
+			);
+			const nodeSchema = definitionMap.get(json[typeField]);
+			assert(nodeSchema?.kind === NodeKind.Object, "Expected object schema");
+		}
+	}
+}
+
+function getSchemaIdentifier(content: TreeEditValue): string | undefined {
+	switch (typeof content) {
+		case "boolean": {
+			return booleanSchema.identifier;
+		}
+		case "number": {
+			return numberSchema.identifier;
+		}
+		case "string": {
+			return stringSchema.identifier;
+		}
+		case "object": {
+			if (content === null) {
+				return nullSchema.identifier;
+			}
+			if (Array.isArray(content)) {
+				throw new UsageError("Arrays are not currently supported in this context");
+			}
+			if (isFluidHandle(content)) {
+				return handleSchema.identifier;
+			}
+			return content[typeField];
+		}
+		default: {
+			throw new UsageError("Unsupported content type");
+		}
+	}
+}
+
+function contentWithIds(content: TreeNode, idGenerator: IdGenerator): TreeEditObject {
+	return JSON.parse(toDecoratedJson(idGenerator, content)) as TreeEditObject;
+}
+
+/**
+ * Manages applying the various types of {@link TreeEdit}'s to a a given {@link TreeNode}.
+ */
+export function applyAgentEdit(
+	treeNode: TreeNode,
+	treeEdit: TreeEdit,
+	idGenerator: IdGenerator,
+	definitionMap: ReadonlyMap<string, SimpleNodeSchema>,
+	validator?: (edit: TreeNode) => void,
+): TreeEdit {
```

Comment:
Nit: this function is very large, and difficult to read. It might make things a bit easier to parse if you parsed out the code for each switch case into a separate function.

Reply:
Agreed, will follow up on the refactor here. I wanted to leave the structure as is to help SharedTree team's review by keeping the code structure familiar.